api/handlers.go

```diff
@@ -111,24 +129,102 @@ func (api *API) userHandlerPUT(w http.ResponseWriter, req *http.Request, ps http
 }
 
 // userChangePasswordHandler changes a user's password, given the old one is known.
-func (api *API) userChangePasswordHandler(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
-	// TODO Implement
-	WriteJSON(w, struct{ msg string }{"Not implemented."})
+func (api *API) userChangePasswordHandler(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	uid, _, _, err := jwtToken(req)
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	var u *database.User
+	// Fetch the user by their id.
+	u, err = api.staticDB.UserByID(req.Context(), uid)
+	if err != nil {
+		// This is a Bad Request and not an Internal Server Error because
+		// the client has supplied an invalid user id.
+		WriteError(w, errors.AddContext(err, "failed to fetch user"), http.StatusBadRequest)
+		return
+	}
+	err = req.ParseMultipartForm(MaxMultipartMem)
+	if err != nil {
+		WriteError(w, errors.New("Failed to parse multipart parameters."), http.StatusBadRequest)
+		return
+	}
+	oldPass := req.PostFormValue("oldPassword")
+	newPass := req.PostFormValue("newPassword")
+	if oldPass == "" || newPass == "" {
+		WriteError(w, errors.New("Both `oldPassword` and `newPassword` are required."), http.StatusBadRequest)
+		return
+	}
+	// Validate that the given old password is correct.
+	err = u.VerifyPassword(oldPass)
+	if err != nil {
+		WriteError(w, errors.New("Bad username or password."), http.StatusBadRequest)
+		return
+	}
+	// Set the new password.
+	err = u.SetPassword(newPass)
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	// Persist the change.
+	err = api.staticDB.UserUpdatePassword(req.Context(), u)
```

Comment:
I think changing the password should be a single method on the database. 
We should never need to fetch a user from a databse at all actually. Unless we explicitly want to return user information.

e.g. if you want to update a user's pw you should call `staticDB.ChangePassword(id, oldPassword, newPassword)`
Then you would (within a single db txn) fetch the user, verify the password and set the password. 

Otherwise you end up with race conditions where 2 threads fetch a user, each one sets a different new password, both verify the old password and then they call update and whoever calls update last wins.

Reply:
tldr; 

Updates to the database should be atomic and that should be reflected in its interface.

Reply:
Hm, this should also extend to "update user data" then. I'll update it.

Reply:
I'm actually not sure about "update user data" - that might have more complicated logic that definitely doesn't fit into a single DB txn.

For the moment I would hold off on adding transactions, so we can get this off the ground. I've added `TODO`.

Reply:
> We should never need to fetch a user from a databse at all actually. Unless we explicitly want to return user information.

We need to fetch the user in order to get their personal salt. Otherwise we cannot verify that the password is correct. I don't think we can do `bcrypt` stuff within Mongo requests.

> Otherwise you end up with race conditions where 2 threads fetch a user, each one sets a different new password, both verify the old password and then they call update and whoever calls update last wins.

We don't have race conditions because we don't have threads for the moment. That's why we removed the mutex from the user struct. We have *some form* of multi-threading in the form of multiple requests that work with the same user record, though.

The only way to avoid the race conditions is to lock/unlock the record in the database or to use transactions. The second approach is what we should do, IMO. But I would argue (as I did above) that we can skip that for the moment and do it in a follow up. If you think that this is required then I'll add transactions.